From e3eb8289ce0cf331132d8279ac80345a1a0c2248 Mon Sep 17 00:00:00 2001
From: Shen-Ta Hsieh <beststeve@secondstate.io>
Date: Fri, 3 Nov 2023 15:29:19 +0800
Subject: [PATCH] [AOT] Add support for LLVM 17

* Use pipeline description string for llvm 13 and above.
* Add definition for LLD 17

Signed-off-by: Shen-Ta Hsieh <beststeve@secondstate.io>

Upstream-Status: Backport [https://github.com/WasmEdge/WasmEdge/commit/e3eb8289ce0cf331132d8279ac80345a1a0c2248]

Adapt the include hunk to WasmEdge 0.13.4, which predates the MSVC macro.
---
 lib/aot/compiler.cpp |  55 +++++++++++++--
 lib/aot/llvm.h       | 160 ++++++++++++++++++++++++++++++++++++++++++-
 2 files changed, 207 insertions(+), 8 deletions(-)

diff --git a/lib/aot/compiler.cpp b/lib/aot/compiler.cpp
index 1f4d9b5db096..cf49c2abba77 100644
--- a/lib/aot/compiler.cpp
+++ b/lib/aot/compiler.cpp
@@ -29,6 +29,15 @@
 #if LLVM_VERSION_MAJOR >= 14
 #include <lld/Common/CommonLinkerContext.h>
 #endif
+#if LLVM_VERSION_MAJOR >= 17
+#if WASMEDGE_OS_MACOS
+LLD_HAS_DRIVER(macho)
+#elif WASMEDGE_OS_LINUX
+LLD_HAS_DRIVER(elf)
+#elif WASMEDGE_OS_WINDOWS
+LLD_HAS_DRIVER(coff)
+#endif
+#endif
 
 #if WASMEDGE_OS_MACOS
 #include <sys/utsname.h>
@@ -136,6 +145,28 @@ static inline constexpr const bool kForceDivCheck = true;
 static inline constexpr const uint32_t kValSize = sizeof(WasmEdge::ValVariant);
 
 // Translate Compiler::OptimizationLevel to llvm::PassBuilder version
+#if LLVM_VERSION_MAJOR >= 13
+static inline const char *
+toLLVMLevel(WasmEdge::CompilerConfigure::OptimizationLevel Level) noexcept {
+  using OL = WasmEdge::CompilerConfigure::OptimizationLevel;
+  switch (Level) {
+  case OL::O0:
+    return "default<O0>,function(tailcallelim)";
+  case OL::O1:
+    return "default<O1>";
+  case OL::O2:
+    return "default<O2>";
+  case OL::O3:
+    return "default<O3>";
+  case OL::Os:
+    return "default<Os>";
+  case OL::Oz:
+    return "default<Oz>";
+  default:
+    assumingUnreachable();
+  }
+}
+#else
 static inline std::pair<unsigned int, unsigned int>
 toLLVMLevel(WasmEdge::CompilerConfigure::OptimizationLevel Level) noexcept {
   using OL = WasmEdge::CompilerConfigure::OptimizationLevel;
@@ -156,6 +187,8 @@ toLLVMLevel(WasmEdge::CompilerConfigure::OptimizationLevel Level) noexcept {
     assumingUnreachable();
   }
 }
+#endif
+
 static inline LLVMCodeGenOptLevel toLLVMCodeGenLevel(
     WasmEdge::CompilerConfigure::OptimizationLevel Level) noexcept {
   using OL = WasmEdge::CompilerConfigure::OptimizationLevel;
@@ -5307,17 +5340,24 @@ Expect<void> Compiler::compile(Span<const Byte> Data, const AST::Module &Module,
           LLVMRelocPIC, LLVMCodeModelDefault);
     }
 
+#if LLVM_VERSION_MAJOR >= 13
+    {
+      auto PBO = LLVM::PassBuilderOptions::create();
+      LLVM::Error Error = PBO.runPasses(
+          LLModule,
+          toLLVMLevel(Conf.getCompilerConfigure().getOptimizationLevel()), TM);
+      if (Error) {
+        spdlog::error("{}"sv, Error.message().string_view());
+      }
+    }
+#else
     {
       auto FP = LLVM::PassManager::createForModule(LLModule);
       auto MP = LLVM::PassManager::create();
 
       TM.addAnalysisPasses(MP);
       TM.addAnalysisPasses(FP);
-      if (Conf.getCompilerConfigure().getOptimizationLevel() ==
-          CompilerConfigure::OptimizationLevel::O0) {
-        FP.addTailCallEliminationPass();
-        MP.addAlwaysInlinerPass();
-      } else {
+      {
         auto PMB = LLVM::PassManagerBuilder::create();
         auto [OptLevel, SizeLevel] =
             toLLVMLevel(Conf.getCompilerConfigure().getOptimizationLevel());
@@ -5326,6 +5366,10 @@ Expect<void> Compiler::compile(Span<const Byte> Data, const AST::Module &Module,
         PMB.populateFunctionPassManager(FP);
         PMB.populateModulePassManager(MP);
       }
+      if (Conf.getCompilerConfigure().getOptimizationLevel() ==
+          CompilerConfigure::OptimizationLevel::O0) {
+        FP.addTailCallEliminationPass();
+      }
 
       FP.initializeFunctionPassManager();
       for (auto Fn = LLModule.getFirstFunction(); Fn;
@@ -5335,6 +5379,7 @@ Expect<void> Compiler::compile(Span<const Byte> Data, const AST::Module &Module,
       FP.finalizeFunctionPassManager();
       MP.runPassManager(LLModule);
     }
+#endif
 
     // Set initializer for constant value
     if (auto IntrinsicsTable = LLModule.getNamedGlobal("intrinsics")) {
diff --git a/lib/aot/llvm.h b/lib/aot/llvm.h
index c393c45ff627..d7179f29019c 100644
--- a/lib/aot/llvm.h
+++ b/lib/aot/llvm.h
@@ -5,14 +5,19 @@
 #include "common/span.h"
 #include <llvm-c/Analysis.h>
 #include <llvm-c/Core.h>
+#include <llvm-c/Error.h>
 #include <llvm-c/Object.h>
 #include <llvm-c/Target.h>
 #include <llvm-c/TargetMachine.h>
+#if LLVM_VERSION_MAJOR >= 13
+#include <llvm-c/Transforms/PassBuilder.h>
+#else
 #include <llvm-c/Transforms/IPO.h>
 #include <llvm-c/Transforms/PassManagerBuilder.h>
 #include <llvm-c/Transforms/Scalar.h>
+#endif
 #include <llvm-c/Types.h>
 #include <mutex>
 #include <string_view>
 #include <utility>
 #include <vector>
@@ -93,7 +99,9 @@ class Core {
   static inline std::once_flag Once;
   static inline void initOnce() noexcept {
     using namespace std::literals;
+#if LLVM_VERSION_MAJOR < 17
     LLVMInitializeCore(LLVMGetGlobalPassRegistry());
+#endif
     LLVMInitializeNativeTarget();
     LLVMInitializeNativeAsmPrinter();
 
@@ -288,6 +296,64 @@ class Module {
   LLVMModuleRef Ref = nullptr;
 };
 
+class ErrorMessage {
+public:
+  constexpr ErrorMessage() noexcept = default;
+  constexpr ErrorMessage(char *M) noexcept : Data(M) {}
+  ErrorMessage(const ErrorMessage &) = delete;
+  ErrorMessage &operator=(const ErrorMessage &) = delete;
+  ErrorMessage(ErrorMessage &&M) noexcept : ErrorMessage() { swap(*this, M); }
+  ErrorMessage &operator=(ErrorMessage &&M) noexcept {
+    swap(*this, M);
+    return *this;
+  }
+
+  ~ErrorMessage() noexcept { LLVMDisposeErrorMessage(Data); }
+
+  constexpr operator bool() const noexcept { return Data != nullptr; }
+  constexpr auto &unwrap() const noexcept { return Data; }
+  constexpr auto &unwrap() noexcept { return Data; }
+  friend void swap(ErrorMessage &LHS, ErrorMessage &RHS) noexcept {
+    using std::swap;
+    swap(LHS.Data, RHS.Data);
+  }
+
+  std::string_view string_view() const noexcept { return Data; }
+
+private:
+  char *Data = nullptr;
+};
+
+class Error {
+public:
+  constexpr Error() noexcept = default;
+  constexpr Error(LLVMErrorRef R) noexcept : Ref(R) {}
+  Error(const Error &) = delete;
+  Error &operator=(const Error &) = delete;
+  Error(Error &&E) noexcept : Error() { swap(*this, E); }
+  Error &operator=(Error &&E) noexcept {
+    swap(*this, E);
+    return *this;
+  }
+
+  ~Error() noexcept { LLVMConsumeError(Ref); }
+
+  constexpr operator bool() const noexcept { return Ref != nullptr; }
+  constexpr auto &unwrap() const noexcept { return Ref; }
+  constexpr auto &unwrap() noexcept { return Ref; }
+  friend void swap(Error &LHS, Error &RHS) noexcept {
+    using std::swap;
+    swap(LHS.Ref, RHS.Ref);
+  }
+
+  ErrorMessage message() noexcept {
+    return LLVMGetErrorMessage(std::exchange(Ref, nullptr));
+  }
+
+private:
+  LLVMErrorRef Ref = nullptr;
+};
+
 class Message {
 public:
   constexpr Message() noexcept = default;
@@ -1509,6 +1575,7 @@ class Target {
   LLVMTargetRef Ref = nullptr;
 };
 
+#if LLVM_VERSION_MAJOR < 13
 class PassManager {
 public:
   constexpr PassManager() noexcept = default;
@@ -1602,6 +1669,7 @@ class PassManagerBuilder {
 private:
   LLVMPassManagerBuilderRef Ref = nullptr;
 };
+#endif
 
 class TargetMachine {
 public:
@@ -1635,9 +1703,11 @@ class TargetMachine {
                                    Reloc, CodeModel);
   }
 
+#if LLVM_VERSION_MAJOR < 13
   void addAnalysisPasses(PassManager &P) noexcept {
     LLVMAddAnalysisPasses(Ref, P.unwrap());
   }
+#endif
 
   std::pair<MemoryBuffer, Message>
   emitToMemoryBuffer(Module &M, LLVMCodeGenFileType CodeGen) noexcept {
@@ -1652,6 +1722,90 @@ class TargetMachine {
   LLVMTargetMachineRef Ref = nullptr;
 };
 
+#if LLVM_VERSION_MAJOR >= 13
+class PassBuilderOptions {
+public:
+  constexpr PassBuilderOptions() noexcept = default;
+  constexpr PassBuilderOptions(LLVMPassBuilderOptionsRef R) noexcept : Ref(R) {}
+  PassBuilderOptions(const PassBuilderOptions &) = delete;
+  PassBuilderOptions &operator=(const PassBuilderOptions &) = delete;
+  PassBuilderOptions(PassBuilderOptions &&O) noexcept : PassBuilderOptions() {
+    swap(*this, O);
+  }
+  PassBuilderOptions &operator=(PassBuilderOptions &&O) noexcept {
+    swap(*this, O);
+    return *this;
+  }
+
+  ~PassBuilderOptions() noexcept { LLVMDisposePassBuilderOptions(Ref); }
+
+  constexpr operator bool() const noexcept { return Ref != nullptr; }
+  constexpr auto &unwrap() noexcept { return Ref; }
+  friend void swap(PassBuilderOptions &LHS, PassBuilderOptions &RHS) noexcept {
+    using std::swap;
+    swap(LHS.Ref, RHS.Ref);
+  }
+
+  static PassBuilderOptions create() noexcept {
+    return LLVMCreatePassBuilderOptions();
+  }
+
+  void setVerifyEach(bool VerifyEach) noexcept {
+    LLVMPassBuilderOptionsSetVerifyEach(Ref, VerifyEach);
+  }
+
+  void setDebugLogging(bool DebugLogging) noexcept {
+    LLVMPassBuilderOptionsSetDebugLogging(Ref, DebugLogging);
+  }
+
+  void setLoopInterleaving(bool LoopInterleaving) noexcept {
+    LLVMPassBuilderOptionsSetLoopInterleaving(Ref, LoopInterleaving);
+  }
+
+  void setLoopVectorization(bool LoopVectorization) noexcept {
+    LLVMPassBuilderOptionsSetLoopVectorization(Ref, LoopVectorization);
+  }
+
+  void setSLPVectorization(bool SLPVectorization) noexcept {
+    LLVMPassBuilderOptionsSetSLPVectorization(Ref, SLPVectorization);
+  }
+
+  void setLoopUnrolling(bool LoopUnrolling) noexcept {
+    LLVMPassBuilderOptionsSetLoopUnrolling(Ref, LoopUnrolling);
+  }
+
+  void setForgetAllSCEVInLoopUnroll(bool ForgetAllSCEVInLoopUnroll) noexcept {
+    LLVMPassBuilderOptionsSetForgetAllSCEVInLoopUnroll(
+        Ref, ForgetAllSCEVInLoopUnroll);
+  }
+
+  void setLicmMssaOptCap(unsigned int LicmMssaOptCap) noexcept {
+    LLVMPassBuilderOptionsSetLicmMssaOptCap(Ref, LicmMssaOptCap);
+  }
+
+  void setLicmMssaNoAccForPromotionCap(
+      unsigned int LicmMssaNoAccForPromotionCap) noexcept {
+    LLVMPassBuilderOptionsSetLicmMssaNoAccForPromotionCap(
+        Ref, LicmMssaNoAccForPromotionCap);
+  }
+
+  void setCallGraphProfile(bool CallGraphProfile) noexcept {
+    LLVMPassBuilderOptionsSetCallGraphProfile(Ref, CallGraphProfile);
+  }
+
+  void setMergeFunctions(bool MergeFunctions) noexcept {
+    LLVMPassBuilderOptionsSetMergeFunctions(Ref, MergeFunctions);
+  }
+
+  Error runPasses(Module &M, const char *Passes, TargetMachine &TM) noexcept {
+    return LLVMRunPasses(M.unwrap(), Passes, TM.unwrap(), Ref);
+  }
+
+private:
+  LLVMPassBuilderOptionsRef Ref = nullptr;
+};
+#endif
+
 class SectionIterator {
 public:
   constexpr SectionIterator() noexcept = default;
